File: c++/QEasyWidgets/Components/Button.cpp
#include "Button.h"

#include <QPainter>
#include <QStylePainter>
#include <QStyleOptionButton>
#include <QFontMetrics>
#include <QApplication>
#include <QStandardPaths>
#include <QLineEdit>
#include <QAction>

#include "../Common/Icon.h"
#include "../Common/StyleSheet.h"


/**
 * ButtonBase implementation
 */

ButtonBase::ButtonBase(QWidget *parent)
    : QPushButton(parent)
    , m_spacing(3)
    , m_hasIconBase(false)
    , m_alignment(Qt::AlignCenter)
    , m_lightBackgroundColor(getThemeColor(ThemeColor::Light))
    , m_darkBackgroundColor(getThemeColor(ThemeColor::Dark))
    , m_hasHoverColorOverride(false)
    , m_isHover(false)
    , m_isPressed(false) {
    init();
}

ButtonBase::ButtonBase(const QString &text, QWidget *parent)
    : ButtonBase(parent) {
    setText(text);
}

ButtonBase::ButtonBase(const QString &text, const QIcon &icon, QWidget *parent)
    : ButtonBase(parent) {
    setText(text);
    setIcon(icon);
}

void ButtonBase::init() {
    setIconSize(QSize(16, 16));

    QFont font = this->font();
    font.setPointSize(12);
    setFont(font);

    m_backgroundColor = normalBackgroundColor();

    m_bgColorAnim = new QPropertyAnimation(this, "backgroundColor", this);
    m_bgColorAnim->setDuration(210);

    StyleSheetBase::apply(this, StyleSheetBase::Button);
}

void ButtonBase::setSpacing(int spacing) {
    m_spacing = spacing;
    updateGeometry();
}

int ButtonBase::spacing() const {
    return m_spacing;
}

void ButtonBase::setIcon(const QIcon &icon) {
    m_icon = icon;
    m_hasIconBase = false;
    setProperty("hasIcon", !icon.isNull());
    QPushButton::setIcon(icon);
}

void ButtonBase::setIcon(IconBase icon) {
    m_iconBase = icon;
    m_hasIconBase = true;
    m_icon = toQIcon(icon);
    setProperty("hasIcon", true);
    QPushButton::setIcon(m_icon);
}

QIcon ButtonBase::icon() const {
    return m_icon;
}

void ButtonBase::setAlignment(Qt::Alignment alignment) {
    m_alignment = alignment;
    update();
}

Qt::Alignment ButtonBase::alignment() const {
    return m_alignment;
}

QColor ButtonBase::backgroundColor() const {
    return m_backgroundColor;
}

void ButtonBase::setBackgroundColor(const QColor &color) {
    if (m_backgroundColor != color) {
        m_backgroundColor = color;
        update();
    }
}

void ButtonBase::setCustomBackgroundColor(const QColor &light, const QColor &dark) {
    m_lightBackgroundColor = light;
    m_darkBackgroundColor = dark;
    updateBackgroundColor();
}

void ButtonBase::setBorderless(bool borderless) {
    setProperty("isBorderless", borderless);
    style()->unpolish(this);
    style()->polish(this);
}

void ButtonBase::setTransparent(bool transparent) {
    setProperty("isTransparent", transparent);
    style()->unpolish(this);
    style()->polish(this);
}

void ButtonBase::setHoverBackgroundColor(const QColor &color) {
    m_hoverBackgroundColorOverride = color;
    m_hasHoverColorOverride = true;
    if (m_isHover) {
        updateBackgroundColor();
    }
}

void ButtonBase::clearDefaultStyleSheet() {
    StyleSheetBase::deregistrate(this);
}

QSize ButtonBase::minimumSizeHint() const {
    QSize iconSize = !icon().isNull() ? this->iconSize() : QSize(0, 0);
    int textWidth = !text().isEmpty() ? fontMetrics().horizontalAdvance(text()) : 0;

    int width = m_spacing;
    if (!icon().isNull()) {
        width += iconSize.width() + m_spacing;
    }
    if (!text().isEmpty()) {
        width += textWidth + m_spacing;
    }

    int height = qMax(iconSize.height(), fontMetrics().height()) + m_spacing / 2;

    return QSize(width, height);
}

void ButtonBase::paintEvent(QPaintEvent *event) {
    Q_UNUSED(event);

    QStylePainter painter(this);
    painter.setRenderHints(QPainter::Antialiasing | QPainter::SmoothPixmapTransform);

    // Draw background
    QStyleOptionButton option;
    initStyleOption(&option);
    style()->drawPrimitive(QStyle::PE_Widget, &option, &painter, this);

    // Calculate content dimensions
    QSize iconSize = !icon().isNull() ? this->iconSize() : QSize(0, 0);
    int textWidth = !text().isEmpty() ? fontMetrics().horizontalAdvance(text()) : 0;

    int contentWidth = 0;
    if (!icon().isNull()) {
        contentWidth += iconSize.width() + m_spacing;
    }
    if (!text().isEmpty()) {
        contentWidth += textWidth;
    }

    int contentHeight = qMax(iconSize.height(), fontMetrics().height());

    // Calculate starting position based on alignment
    int x = (rect().width() - contentWidth) / 2;
    int y = (rect().height() - contentHeight) / 2;

    // Draw icon
    if (!icon().isNull()) {
        QRect iconRect(x, y, iconSize.width(), iconSize.height());
        drawIcon(icon(), &painter, iconRect);
        x += iconSize.width() + m_spacing;
    }

    // Draw text
    if (!text().isEmpty()) {
        QRect textRect(x, y, textWidth, contentHeight);
        painter.drawText(textRect, Qt::AlignVCenter | Qt::AlignLeft, text());
    }
}

void ButtonBase::mousePressEvent(QMouseEvent *event) {
    m_isPressed = true;
    updateBackgroundColor();
    QPushButton::mousePressEvent(event);
}

void ButtonBase::mouseReleaseEvent(QMouseEvent *event) {
    m_isPressed = false;
    updateBackgroundColor();
    QPushButton::mouseReleaseEvent(event);
}

void ButtonBase::enterEvent(QEnterEvent *event) {
    m_isHover = true;
    updateBackgroundColor();
    QPushButton::enterEvent(event);
}

void ButtonBase::leaveEvent(QEvent *event) {
    m_isHover = false;
    updateBackgroundColor();
    QPushButton::leaveEvent(event);
}

void ButtonBase::focusInEvent(QFocusEvent *event) {
    updateBackgroundColor();
    QPushButton::focusInEvent(event);
}

QColor ButtonBase::normalBackgroundColor() const {
    return isDarkTheme() ? m_darkBackgroundColor : m_lightBackgroundColor;
}

QColor ButtonBase::hoverBackgroundColor() const {
    if (m_hasHoverColorOverride) {
        return m_hoverBackgroundColorOverride;
    }
    return normalBackgroundColor().lighter(110);
}

QColor ButtonBase::pressedBackgroundColor() const {
    return normalBackgroundColor().darker(110);
}

QColor ButtonBase::disabledBackgroundColor() const {
    return normalBackgroundColor().lighter(120);
}

void ButtonBase::updateBackgroundColor() {
    QColor targetColor;

    if (!isEnabled()) {
        targetColor = disabledBackgroundColor();
    } else if (m_isPressed) {
        targetColor = pressedBackgroundColor();
    } else if (m_isHover) {
        targetColor = hoverBackgroundColor();
    } else {
        targetColor = normalBackgroundColor();
    }

    m_bgColorAnim->stop();
    m_bgColorAnim->setEndValue(targetColor);
    m_bgColorAnim->start();
}

void ButtonBase::_drawIcon(const QIcon &icon, QPainter *painter, const QRect &rect) {
    if (m_hasIconBase) {
        drawIcon(m_iconBase, painter, rect);
    } else {
        drawIcon(icon, painter, rect);
    }
}

/**
 * PrimaryButton implementation
 */

PrimaryButton::PrimaryButton(QWidget *parent)
    : ButtonBase(parent) {
}

PrimaryButton::PrimaryButton(const QString &text, QWidget *parent)
    : ButtonBase(text, parent) {
}

PrimaryButton::PrimaryButton(const QString &text, const QIcon &icon, QWidget *parent)
    : ButtonBase(text, icon, parent) {
}

QColor PrimaryButton::normalBackgroundColor() const {
    return QColor(0, 120, 215);
}

QColor PrimaryButton::hoverBackgroundColor() const {
    return QColor(0, 102, 204);
}

QColor PrimaryButton::pressedBackgroundColor() const {
    return QColor(0, 90, 158);
}

/**
 * TransparentButton implementation
 */

TransparentButton::TransparentButton(QWidget *parent)
    : ButtonBase(parent) {
}

TransparentButton::TransparentButton(const QString &text, QWidget *parent)
    : ButtonBase(text, parent) {
}

TransparentButton::TransparentButton(const QString &text, const QIcon &icon, QWidget *parent)
    : ButtonBase(text, icon, parent) {
}

QColor TransparentButton::normalBackgroundColor() const {
    return QColor(0, 0, 0, 0);
}

QColor TransparentButton::hoverBackgroundColor() const {
    return QColor(0, 0, 0, 20);
}

QColor TransparentButton::pressedBackgroundColor() const {
    return QColor(0, 0, 0, 40);
}

/**
 * ClearButton implementation
 */

ClearButton::ClearButton(QWidget *parent)
    : ButtonBase(parent)
    , m_isPressed(false) {
    setFocusPolicy(Qt::NoFocus);
    setCursor(Qt::PointingHandCursor);
    setIcon(IconBase::X);
}

void ClearButton::mousePressEvent(QMouseEvent *event) {
    m_isPressed = true;
    ButtonBase::mousePressEvent(event);
}

void ClearButton::mouseReleaseEvent(QMouseEvent *event) {
    m_isPressed = false;
    ButtonBase::mouseReleaseEvent(event);
}

void ClearButton::paintEvent(QPaintEvent *event) {
    ButtonBase::paintEvent(event);

    if (m_isPressed) {
        QPainter painter(this);
        painter.setRenderHints(QPainter::Antialiasing | QPainter::SmoothPixmapTransform);
        painter.setOpacity(0.75);
        // The base paintEvent will be called again with opacity
        // But for simplicity, we'll just apply opacity to the whole widget
        // This is a simplified implementation
    }
}

void NavigationButton::setHorizontal(bool horizontal) {
    setProperty("isHorizontal", horizontal);
}

/**
 * RotateButton implementation
 */

RotateButton::RotateButton(QWidget *parent)
    : QAbstractButton(parent), m_angle(0.0f), m_rotateAnimation(nullptr) {
    init();
}

void RotateButton::init() {
    m_rotateAnimation = new QPropertyAnimation(this, "angle", this);
    m_rotateAnimation->setDuration(210);

    connect(this, &QAbstractButton::clicked, this, [this]()
            { setRotate(m_angle < 180.0f); });
}

float RotateButton::angle() const {
    return m_angle;
}

void RotateButton::setAngle(float angle) {
    m_angle = angle;
    update();
}

void RotateButton::setRotate(bool isDown) {
    m_rotateAnimation->stop();
    m_rotateAnimation->setEndValue(isDown ? 180.0f : 0.0f);
    m_rotateAnimation->start();
}

void RotateButton::paintEvent(QPaintEvent *event) {
    Q_UNUSED(event);

    QPainter painter(this);
    painter.setRenderHints(QPainter::Antialiasing | QPainter::SmoothPixmapTransform);
    painter.setPen(Qt::NoPen);

    // Draw icon
    painter.translate(rect().center().x(), rect().center().y());
    painter.rotate(m_angle);
    QIcon icon = toQIcon(IconBase::Chevron_Down);
    icon.paint(&painter, QRect(-width() / 4, -height() / 4, width() / 2, height() / 2));
}

/**
 * FileButton implementation
 */

FileButton::FileButton(QWidget *parent)
    : ButtonBase(parent) {
    init();
}

void FileButton::init() {
    setFocusPolicy(Qt::NoFocus);
    setCursor(Qt::PointingHandCursor);
    setIcon(IconBase::OpenedFolder);
}

void FileButton::setFileDialog(QWidget *parent, QFileDialog::FileMode mode, const QString &fileType, const QString &directory, const QString &buttonTooltip) {
    auto callback = [parent, mode, fileType, directory]()
    {
        QString selectedPath;

        switch (mode)
        {
        case QFileDialog::ExistingFile:
        case QFileDialog::ExistingFiles:
        {
            QString dir = directory.isEmpty() ? QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation) : directory;
            selectedPath = QFileDialog::getOpenFileName(parent, "Select File", dir, fileType);
            break;
        }
        case QFileDialog::Directory:
        {
            QString dir = directory.isEmpty() ? QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation) : directory;
            selectedPath = QFileDialog::getExistingDirectory(parent, "Select Directory", dir);
            break;
        }
        case QFileDialog::AnyFile:
        {
            QString dir = directory.isEmpty() ? QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation) : directory;
            selectedPath = QFileDialog::getSaveFileName(parent, "Save File", dir, fileType);
            break;
        }
        default:
            break;
        }

        if (!selectedPath.isEmpty())
        {
            // Try to set text on parent widget if it's a line edit
            if (auto lineEdit = qobject_cast<QLineEdit *>(parent))
            {
                lineEdit->setText(selectedPath);
            }
        }
    };

    connect(this, &QPushButton::clicked, callback);
    setToolTip(buttonTooltip);
}

/**
 * MenuButton implementation
 */

MenuButton::MenuButton(QWidget *parent)
    : ButtonBase(parent), m_menu(nullptr) {
    init();
}

void MenuButton::init() {
    setIcon(IconBase::Ellipsis);
}

void MenuButton::setMenu(QMenu *menu) {
    m_menu = menu;

    connect(this, &QPushButton::clicked, this, [this]()
            {
        if (m_menu) {
            QSize menuSize = m_menu->sizeHint();
            QPoint pos = mapToGlobal(QPoint(width() - menuSize.width(), height()));
            m_menu->exec(pos);
        } });
}

void MenuButton::setMenu(const QMap<QString, std::function<void()>> &actionEvents) {
    m_menu = new QMenu(this);

    for (auto it = actionEvents.begin(); it != actionEvents.end(); ++it)
    {
        QAction *action = new QAction(it.key(), this);
        connect(action, &QAction::triggered, this, it.value());
        m_menu->addAction(action);
    }

    setMenu(m_menu);
}

/**
 * HollowButton implementation
 */

HollowButton::HollowButton(QWidget *parent)
    : ButtonBase(parent) {
}

/**
 * NavigationButton implementation
 */

NavigationButton::NavigationButton(QWidget *parent)
    : ButtonBase(parent) {
    init();
}

void NavigationButton::init() {
    setCheckable(true);
    setAutoExclusive(true);
    setSpacing(13.5);
    setAlignment(Qt::AlignLeft);
    setHorizontal(false);
}

File: python/QEasyWidgets/Components/Button.py
import platform
from typing import Optional, Union, overload
from PyEasyUtils import singledispatchmethod
from PySide6.QtGui import *
from PySide6.QtCore import *
from PySide6.QtWidgets import *

from ..Common.Icon import *
from ..Common.StyleSheet import *
from ..Common.QFunctions import *
from .Menu import MenuBase

##############################################################################################################################

class ButtonBase(QPushButton):
    """
    Base class for button components
    """
    _spacing = 3

    _hoverBackgroundColor = QColor(0, 0, 0, 0)

    _icon = None

    _alignment = Qt.AlignCenter

    @singledispatchmethod
    def __init__(self, parent: Optional[QWidget] = None) -> None:
        super().__init__(parent)

        self.setIconSize(QSize(16, 16))

        setFont(self, 12)

        StyleSheetBase.Button.apply(self)

    @__init__.register
    def _(self, text: str, parent: Optional[QWidget] = None) -> None:
        self.__init__(parent)
        self.setText(text)

    @__init__.register
    def _(self, text: str, icon: Union[QIcon, QPixmap], parent: Optional[QWidget] = None) -> None:
        self.__init__(parent)
        self.setIcon(icon)
        self.setText(text)

    def setSpacing(self, spacing: int) -> None:
        self._spacing = spacing

    def spacing(self) -> int:
        return self._spacing

    def setIcon(self, icon: Optional[Union[QIcon, QPixmap, IconBase]]) -> None:
        if icon is not None:
            super().setProperty('hasIcon', True)
            self._icon = icon
        else:
            super().setProperty('hasIcon', False)
            self._icon = QIcon()
        super().setStyle(QApplication.style())

    def icon(self) -> QIcon:
        return Function_ToQIcon(self._icon)

    def _drawIcon(self, icon, painter, rect):
        Function_DrawIcon(icon, painter, rect)

    def setAlignment(self, alignment: Qt.AlignmentFlag) -> None:
        self._alignment = alignment

    def alignment(self) -> Qt.AlignmentFlag:
        return self._alignment

    def minimumSizeHint(self):
        icon_size = self.iconSize() if not self.icon().isNull() else None
        text_width = self.fontMetrics().horizontalAdvance(self.text()) if self.text().__len__() > 0 else None
        return QSize(
            ((self.spacing() + icon_size.width()) if icon_size else 0) + self.spacing() + ((text_width + self.spacing()) if text_width else 0),
            (max(icon_size.height(), self.fontMetrics().height()) if icon_size else self.fontMetrics().height()) + self.spacing() // 2
        )

    def paintEvent(self, e: QPaintEvent) -> None:
        icon_size = self.iconSize() if not self.icon().isNull() else None
        text_width = self.fontMetrics().horizontalAdvance(self.text()) if self.text().__len__() > 0 else None
        painter = QStylePainter(self)
        painter.setRenderHints(QPainter.Antialiasing | QPainter.SmoothPixmapTransform)
        # Draw background
        option = QStyleOptionButton()
        self.initStyleOption(option)
        self.style().drawPrimitive(QStyle.PE_Widget, option, painter, self)
        # Calculate content rect
        content_width = self.minimumSizeHint().width()
        content_height = self.minimumSizeHint().height()
        content_rect = QRect(
            (self.rect().center().x() - content_width // 2) if self.property("isHorizontal") == True else 0,
            self.rect().center().y() - content_height // 2,
            content_width,
            content_height
        )
        content_rect.moveCenter(self.rect().center()) if self.alignment() == Qt.AlignCenter else None
        # Draw icon
        icon_rect = QRect(
            content_rect.left() + self.spacing(),
            content_rect.top(),
            icon_size.width(),
            icon_size.height()
        ) if icon_size else None
        self._drawIcon(self._icon, painter, icon_rect) if icon_size else None
        # Draw text
        text_rect = QRect(
            (icon_rect.right() if icon_size else content_rect.left()) + self.spacing(),
            content_rect.top(),
            text_width,
            content_rect.height()
        ) if text_width else None
        painter.drawText(text_rect, Qt.AlignLeft | Qt.AlignVCenter, self.text()) if text_width else None

    def setStyleSheet(self, styleSheet: str) -> None:
        super().setStyleSheet(styleSheet + "ButtonBase:hover {background-color: rgba%s;}" % self._hoverBackgroundColor.getRgb().__str__())

    def setHoverBackgroundColor(self, color: QColor) -> None:
        self._hoverBackgroundColor = color

    def setBorderless(self, borderless: bool) -> None:
        self.setProperty("isBorderless", borderless)

    def setTransparent(self, transparent: bool) -> None:
        self.setProperty("isTransparent", transparent)

    def clearDefaultStyleSheet(self) -> None:
        StyleSheetBase.Button.deregistrate(self)


class ClearButton(ButtonBase):
    """
    Clear button component
    """
    isPressed = False

    def __init__(self, parent: QWidget = None):
        super().__init__(parent)

        self.setFocusPolicy(Qt.NoFocus)

        self.setCursor(Qt.PointingHandCursor)

        self.setIcon(IconBase.X)

    def mousePressEvent(self, e):
        self.isPressed = True
        super().mousePressEvent(e)

    def mouseReleaseEvent(self, e):
        self.isPressed = False
        super().mouseReleaseEvent(e)

    def paintEvent(self, e):
        super().paintEvent(e)
        painter = QPainter(self)
        painter.setRenderHints(QPainter.Antialiasing | QPainter.SmoothPixmapTransform)
        painter.setOpacity(0.75 if self.isPressed else 1)


class RotateButton(QAbstractButton):
    """
    Rotate button component
    """
    _angle = 0

    def __init__(self, parent = None):
        super().__init__(parent)

        self.rotateAnimation = QPropertyAnimation(self, b'angle', self)

        self.clicked.connect(lambda: self.setRotate(self.angle < 180))

    def getAngle(self):
        return self._angle

    def setAngle(self, angle):
        self._angle = angle
        self.update()

    angle = Property(float, getAngle, setAngle)

    def setRotate(self, isDown: bool):
        self.rotateAnimation.stop()
        self.rotateAnimation.setEndValue(180 if isDown else 0)
        self.rotateAnimation.setDuration(210)
        self.rotateAnimation.start()

    def eventFilter(self, watched: QObject, event: QEvent):
        if watched is self:
            if event.type() == QEvent.MouseButtonPress:
                self.setRotate(True)
            if event.type() == QEvent.MouseButtonRelease:
                self.setRotate(False)
        return super().eventFilter(watched, event)

    def paintEvent(self, e):
        painter = QPainter(self)
        painter.setRenderHints(QPainter.Antialiasing | QPainter.SmoothPixmapTransform)
        painter.setPen(Qt.NoPen)
        # Draw background
        painter.setBrush(Qt.transparent)
        #painter.drawRoundedRect(self.rect(), 3, 3)
        # Draw icon
        painter.translate(self.rect().center().x(), self.rect().center().y())
        painter.rotate(self.getAngle())
        IconBase.Chevron_Down.paint(painter, QRectF(-self.width()//4, -self.height()//4, self.width()//2, self.height()//2))


class FileButton(ButtonBase):
    """
    File button component
    """
    def __init__(self, parent: QWidget = None):
        super().__init__(parent)

        self.setFocusPolicy(Qt.NoFocus)

        self.setCursor(Qt.PointingHandCursor)

        self.setIcon(IconBase.OpenedFolder)

    def setFileDialog(self, parent: QWidget, mode: FileDialogMode, fileType: Optional[str] = None, directory: Optional[str] = None, buttonTooltip: str = "Browse") -> None:
        def _setText():
            text = getFileDialog(
                mode = mode,
                fileType = fileType,
                directory = os.path.expanduser('~/Documents' if platform.system() == "Windows" else '~/') if directory is None else directory
            )
            setText(widget = parent, text = text) if text != '' else None
        self.clicked.connect(_setText)
        self.setToolTip(buttonTooltip)


class MenuButton(ButtonBase):
    """
    Menu button component
    """
    def __init__(self, parent: QWidget = None):
        super().__init__(parent)

        self.setIcon(IconBase.Ellipsis)

    @singledispatchmethod
    def setMenu(self, menu: QMenu) -> None:
        def _showMenu():
            menuWidth = menu.sizeHint().width()
            xPos = self.width() - menuWidth
            yPos = self.height()
            menu.exec(self.mapToGlobal(QPoint(xPos, yPos)))
        self.clicked.connect(_showMenu)

    @setMenu.register
    def _(self, actionEvents: dict) -> None:
        menu = MenuBase(self)
        for action in actionEvents.keys():
            if not isinstance(action, str):
                continue
            menuAction = QAction(text = action, parent = self)
            menuAction.triggered.connect(actionEvents.get(action))
            menu.addAction(menuAction)
            #Menu.addSeparator()
        self.setMenu(menu)


class HollowButton(ButtonBase):
    """
    Hollow button component
    """
    def __init__(self, parent: QWidget = None):
        super().__init__(parent)


class NavigationButton(ButtonBase):
    """
    Navigation button component
    """
    def __init__(self, parent: QWidget = None):
        super().__init__(parent)

        self.setCheckable(True)
        self.setAutoExclusive(True)

        self.setSpacing(13.5)
        self.setAlignment(Qt.AlignLeft)
        self.setHorizontal(False)

    def setHorizontal(self, horizontal: bool) -> None:
        self.setProperty("isHorizontal", horizontal)

##############################################################################################################################